### ISSUE TITLE
Sequence Diagram - Search & Filter

### ISSUE THREAD

**Issue by bilalcim** [bounswe/bounswe2022group3#105]
As part of this week's submission, a sequence diagram illustrating the interactions between users, methods, and database will be drawn.

First, the user will search for some course and the system will respond accordingly. This part of the diagram will include
* a user searching for a course,
* the procedures behind searching,
* querying the database,
* returning the results.

Then, the user will filter the results by date, and the corresponding part of the sequence diagram will include
* invocation of the filtering procedure on the results,
* returning the results in the chronological order.

**Deadline:** April 5, 2022.
**Reviewer:** The work will be reviewed in the team meeting at April 5, 2022.

**Comment by bilalcim** [comment]
A sequence diagram with the promised ingredients is drawn as follows:

<img width="613" alt="Screen Shot 2022-04-05 at 19 55 22" src="https://user-images.githubusercontent.com/44259352/161811686-5e1fba89-b6d7-4fc1-9b9b-742c5fab7b1c.png">

The diagram will be reviewed at the meeting of April 5.

**Comment by muhammetssen** [comment]
Hi,
- I could not understand why user has to return all of the List<Courses> back to the system in order to filter by date. They could just provide the user_id and search_key as well as a date interval. 
- Similarly, what does `forward` mean, is it for sorting instead of filtering?

**Comment by bilalcim** [comment]
Hi!

> I could not understand why user has to return all of the List back to the system in order to filter by date. They could just provide the user_id and search_key as well as a date interval.

I have thought of searching and filtering as non-overlapping functions. Sure, if you feed searching into the filtering methods, passing the ``user_id`` and ``search_key`` should do.
> Similarly, what does `forward` mean, is it for sorting instead of filtering?

I figured we might as well order them by date if we are to filter a specific period, but of course this is liable to change. (And ``forward`` is the direction of this sorting where ``forward = True`` means chronological.)

**Comment by bilalcim** [comment]
As per the decisions in the last team meeting, I have made the following changes in the searching sequence diagram:

* Added the life cycle rectangles.
* The return arrows are now dotted.
* Also, I somehow mixed up filtering with sorting; therefore, now I have added an ``orderBy`` parameter to each method. It takes its value from an enum class ``order_options`` under the ``SearchService`` class and decides on the order of the courses in the result accordingly. The enum class should hold some options like ``date``, ``rating``, and so on.

![uml](https://user-images.githubusercontent.com/44259352/162590865-7d116c6e-b487-4f83-aafb-4c7bf4638f25.png)

These adjustments can also be seen at the meeting notes which I will reference when it is available.

**Comment by muhammetssen** [comment]
We have reviewed the updated diagram in a weekly team meeting.

**Comment by muhammetssen** [comment]
We have reviewed the updated diagram in a weekly team meeting.

**Comment by bilalcim** [comment]
I have removed the data types from the sequence diagram because in the customer meeting of April 13 we were told to exclude the data types as they can be inferred from the context. Also, I have changed the actor User to RegisteredUser.

![uml](https://user-images.githubusercontent.com/44259352/163341478-a35f1af7-4419-4bfc-a8ab-d6aedc30a0a2.png)

Since all the changes follow from the meeting word by word, I won't reopen the issue for review.

I will be updating the main post accordingly to keep the important information at the top.
